frontend/src/metabase/parameters/components/Parameters/syncQueryParamsWithURL.js

```diff
@@ -0,0 +1,101 @@
+import Dimension from "metabase-lib/lib/Dimension";
+
+export const syncQueryParamsWithURL = props => {
+  props.commitImmediately
+    ? syncForInternalQuestion(props)
+    : syncForPublicQuestion(props);
+};
+
+const syncForInternalQuestion = props => {
+  const { parameters, setParameterValue, query, metadata } = props;
+
+  if (!setParameterValue) {
+    return;
+  }
+
+  for (const parameter of parameters) {
+    const queryParam = query && query[parameter.slug];
+
+    if (queryParam != null || parameter.default != null) {
+      const parsedParam = parseQueryParams(queryParam, parameter, metadata);
+
+      setParameterValue(parameter.id, parsedParam);
+    }
+  }
+};
+
+const syncForPublicQuestion = props => {
+  const { parameters, setParameterValue, query, metadata } = props;
+
+  if (!setParameterValue) {
+    return;
+  }
+
+  const parameterValues = parameters.reduce((acc, parameter) => {
+    const queryParam = query && query[parameter.slug];
+
+    if (queryParam != null || parameter.default != null) {
+      acc[parameter.id] = parseQueryParams(queryParam, parameter, metadata);
+    }
+
+    return acc;
+  }, {});
+
+  setParameterValue(parameterValues);
```

Comment:
We now pass `setParameterValue`, as defined by this branch of logic, an object with 1 or more items.

The root cause of the issue was that we were calling this function 1 or more times with one set of data each time.
The function updated the state multiple times, and overwrote state — even though it nominally used the previous state.

So now we passed all the items at once.
We do it only for public questions.

I tried to do it for all questions, but the logic for internal questions goes through a completely different route, including redux actions.